java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java

```diff
@@ -0,0 +1,229 @@
+
+  private static final MethodMatchers PROCESS_BUILDER_MATCHER = MethodMatchers.create()
+    .ofTypes("java.lang.ProcessBuilder")
+    .constructor()
+    .withAnyParameters()
+    .build();
+
+  private static final MethodMatchers PROCESS_BUILDER_COMMAND_MATCHER = MethodMatchers.create()
+    .ofTypes("java.lang.ProcessBuilder")
+    .names("command")
+    .withAnyParameters()
+    .build();
+
+  private static final MethodMatchers LIST_CREATION_MATCHER = MethodMatchers.or(
+    MethodMatchers.create()
+      .ofTypes("java.util.Arrays")
+      .names("asList")
+      .withAnyParameters()
+      .build(),
+    MethodMatchers.create()
+      .ofTypes("java.util.Collections")
+      .names("singletonList")
+      .withAnyParameters()
+      .build(),
+    MethodMatchers.create()
+      .ofTypes(LIST_TYPE)
+      .names("of")
+      .withAnyParameters()
+      .build()
+  );
+
+  private static final List<String> STARTS = Arrays.asList(
+    "/",
+    "./",
+    "../",
+    "~/",
+    "\\",
+    ".\\",
+    "..\\"
+  );
+
+  private static final Pattern WINDOWS_DISK_PATTERN = Pattern.compile("^[A-Z]:\\\\.*");
+
+  private static final String MESSAGE = "Make sure the \"PATH\" used to find this command includes only what you intend.";
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return MethodMatchers.or(
+      RUNTIME_EXEC_MATCHER,
+      PROCESS_BUILDER_MATCHER,
+      PROCESS_BUILDER_COMMAND_MATCHER
+    );
+  }
+
+  @Override
+  protected void onConstructorFound(NewClassTree tree) {
+    process(tree.arguments());
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree tree) {
+    process(tree.arguments());
+  }
+
+  private void process(Arguments arguments) {
+    if (arguments.isEmpty()) {
+      return;
+    }
+    ExpressionTree firstArgument = ExpressionUtils.skipParentheses(arguments.get(0));
+    processArgument(firstArgument);
+  }
+
+  private void processArgument(ExpressionTree argument) {
+    switch (argument.kind()) {
+      case STRING_LITERAL:
+        if (!isStringLiteralCommandValid(argument)) {
+          reportIssue(argument, MESSAGE);
+        }
+        break;
+      case NEW_ARRAY:
+        if (!isNewArrayCommandValid((NewArrayTree) argument)) {
+          reportIssue(argument, MESSAGE);
+        }
+        break;
+      case IDENTIFIER:
+        if (!isIdentifierCommandValid((IdentifierTree) argument)) {
+          reportIssue(argument, MESSAGE);
+        }
+        break;
+      case METHOD_INVOCATION:
+        if (!isListCommandValid((MethodInvocationTree) argument)) {
+          reportIssue(argument, MESSAGE);
+        }
+        break;
+      default:
+        break;
+    }
+  }
+
+  private static boolean isCompliant(String command) {
+    return STARTS.stream().anyMatch(command::startsWith) ||
+      WINDOWS_DISK_PATTERN.matcher(command).matches();
+  }
+
+  private static boolean isStringLiteralCommandValid(ExpressionTree stringLiteral) {
```

Comment:
In fact, the argument is not always a "stringLiteral", it can be an identifier aswell. We should name it expression to better reflect that.